From c935a582727c06fe656de7d7ee7f95b52a46d5bc Mon Sep 17 00:00:00 2001
From: PeterN <peter1138@openttd.org>
Date: Sun, 8 Jan 2023 18:09:38 +0000
Subject: [PATCH 0835/1053] Fix #10331: Starting new company during load must
 happen after AI start. (#10332)

This situation occurs when loading a savegame in single-player which only
has AI companies.
---
 src/saveload/afterload.cpp | 27 ++++++++++++++-------------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/src/saveload/afterload.cpp b/src/saveload/afterload.cpp
index dfbfef5def..09de2e80ad 100644
--- a/src/saveload/afterload.cpp
+++ b/src/saveload/afterload.cpp
@@ -828,18 +828,6 @@ bool AfterLoadGame()
 	 * version. It didn't show up before r12070. */
 	if (IsSavegameVersionBefore(SLV_87)) UpdateVoidTiles();
 
-	/* If Load Scenario / New (Scenario) Game is used,
-	 *  a company does not exist yet. So create one here.
-	 * 1 exception: network-games. Those can have 0 companies
-	 *   But this exception is not true for non-dedicated network servers! */
-	if (!_networking || (_networking && _network_server && !_network_dedicated)) {
-		CompanyID first_human_company = GetFirstPlayableCompanyID();
-		if (!Company::IsValidID(first_human_company)) {
-			Company *c = DoStartupNewCompany(false, first_human_company);
-			c->settings = _settings_client.company;
-		}
-	}
-
 	/* Fix the cache for cargo payments. */
 	for (CargoPayment *cp : CargoPayment::Iterate()) {
 		cp->front->cargo_payment = cp;
@@ -3234,9 +3222,22 @@ bool AfterLoadGame()
 
 	AfterLoadLinkGraphs();
 
-	/* Start the scripts. This MUST happen after everything else. */
+	/* Start the scripts. This MUST happen after everything else except
+	 * starting a new company. */
 	StartScripts();
 
+	/* If Load Scenario / New (Scenario) Game is used,
+	 *  a company does not exist yet. So create one here.
+	 * 1 exception: network-games. Those can have 0 companies
+	 *   But this exception is not true for non-dedicated network servers! */
+	if (!_networking || (_networking && _network_server && !_network_dedicated)) {
+		CompanyID first_human_company = GetFirstPlayableCompanyID();
+		if (!Company::IsValidID(first_human_company)) {
+			Company *c = DoStartupNewCompany(false, first_human_company);
+			c->settings = _settings_client.company;
+		}
+	}
+
 	return true;
 }
 
-- 
2.55.0

